core/src/test/java/org/apache/accumulo/core/data/MutationTest.java

```diff
@@ -183,6 +185,190 @@ private Value nv(String s) {
     return new Value(s.getBytes());
   }
 
+  @Test
+  public void testAtFamilyTypes() {
+    final String fam = "f16bc";
+    final String qual = "q1pm2";
+    final String val = "v8672194923750";
+
+    Mutation expected = new Mutation("row5");
+    expected.put(fam, qual, val);
+
+    // Test all family methods, keeping qual and val constant as Strings
+    // fam: byte[]
+    Mutation actual = new Mutation("row5");
+    actual.at().family(fam.getBytes(UTF_8)).qualifier(qual).set(val);
```

Comment:
This makes me uncomfortable seeing Mutation be mutable. Can the fluent API be applied as a MutationBuilder instead of to the Object we're creating? I think that would also remove the need for the (to me) odd `at()` method.

e.g.

`Mutation actual = Mutation.newBuilder("row5").family(fam.getBytes(UTF_8)).qualifier(qual).set(val);`

Reply:
To make `Mutation` immutable we'd have to remove the `put()` methods since those allow us to change the `Mutation`'s `UnsynchronizedBuffer`. We'd also have to remove `ConditionalMutation`'s ability to add conditions. 

The consensus in the JIRA ticket and PR comments seems to be that `row` should be immutable but other aspects are fair game to change after object creation.

Reply:
Oh durr. Sorry, I was thinking about ServerMutation (I think is the name of it). Are you aware of that sub-class?

Is there a reason to not make the fluent-methods you want to add on a new class, e.g. MutationBuilder?

Reply:
> Is there a reason to not make the fluent-methods you want to add on a new class, e.g. MutationBuilder?

I was puzzling over this question and shared what I came up with on the issue.

https://issues.apache.org/jira/browse/ACCUMULO-4746?focusedCommentId=16269423&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16269423

Reply:
Since the goal (in my view at least) is to have the method chain be easy to use semantically for the caller, using a new class could be a way to go. As long as the API call stays streamlined like below and the user doesn't have to necessarily be aware of this new class in order to make modifications on their `Mutation`.

`m.at().family().qualifier().visibility().set()`

So in the case of a new class the `at` method could return a new class instead of the nested interface class it's returning now.

Reply:
> I was puzzling over this question and shared what I came up with on the issue.

Thanks for the context, Keith.

> As long as the API call stays streamlined like below and the user doesn't have to necessarily be aware of this new class in order to make modifications on their Mutation.

Ya, that's what I was thinking. Less junk in one Java file. That has the indirect benefit of preventing unintended changes to the existing public API of Mutation.

Reply:
> Less junk in one Java file.

The new interfaces and implementation may not have to be in the the Mutation file.  The only thing that has to be there is the `at()` method.

Reply:
> That has the indirect benefit of preventing unintended changes to the existing public API of Mutation.

In general we should be using tools to check for API changes between releases.